### PR TITLE
add wallet rpc service

### PR DESCRIPTION
**Pull request by jamesvan2019**

### CODE REVIEW

**Comment by jamesvan2019** [comment]
Fixed these comments

File: .gitignore [hunk 1]
```diff
@@ -28,3 +28,7 @@ cmd/payledger/payledger
 cmd/qitmeerd/qitmeerd
 cmd/rlpdump/rlpdump
 *_checksum.txt
+rpc/client/example/qitmeerwalletwebsocket/qitmeerwalletwebsocket
+
+# Go
+go.sum
```
[the remark this answers is not in the document]
Comment (jamesvan2019) [reply .gitignore:34]: had already resoved

File: rpc/errors.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ func (e *invalidMessageError) ErrorCode() int { return -32700 }
 
 func (e *invalidMessageError) Error() string { return e.message }
 
-// unable to decode supplied params, or an invalid number of parameters
+// unable to decode supplied ChainParams, or an invalid number of parameters
```
[the remark this answers is not in the document]
Comment (jamesvan2019) [reply rpc/errors.go]: resolved

**Review by lochjin — CHANGES_REQUESTED** [review]

File: rpc/watchconfirm.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	}
+	for tx, txconf := range *w {
+		txHash := hash.MustHexToDecodedHash(tx)
+		blockRegion, err := txIndex.TxBlockRegion(txHash)
+		if err != nil {
+			log.Error(err.Error(), "txhash", txHash)
+			continue
+		}
+		if blockRegion == nil {
+			if bc.CacheInvalidTx {
```
Comment (lochjin) [inline rpc/watchconfirm.go:47]: Your purpose is just to confirm the transaction, why not `continue` directly?  I don't think it's necessary to deal with all the illegal transactions.

File: rpc/watchconfirm.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+		// Deserialize the transaction
+		var msgTx types.Transaction
+		err = msgTx.Deserialize(bytes.NewReader(txBytes))
+		log.Trace("GetRawTx", "hex", hex.EncodeToString(txBytes))
+		if err != nil {
+			log.Error("Failed to deserialize transaction")
+			continue
+		}
+		mtx := types.NewTx(&msgTx)
+		mtx.IsDuplicate = bc.IsDuplicateTx(mtx.Hash(), blockRegion.Hash)
```
Comment (lochjin) [inline rpc/watchconfirm.go:73]:
```
if mtx.IsDuplicate {
   continue
}
```
why not `continue` directly? I don't think it's necessary to deal with all the illegal transactions.

**Review by lochjin — CHANGES_REQUESTED** [review]

File: rpc/watchconfirm.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		ib := bc.BlockDAG().GetBlock(blockRegion.Hash)
+		if ib == nil {
+			log.Error("block hash not exist", "hash", blockRegion.Hash)
+			delete(*w, tx)
+			continue
+		}
+		confirmations := bc.BlockDAG().GetConfirmations(ib.GetID())
+		isBlue := true
+		if mtx.Tx.IsCoinBase() {
+			isBlue = bc.BlockDAG().IsBlue(ib.GetID())
```
Comment (lochjin) [inline rpc/watchconfirm.go:83]:
```
if isBlue {
   continue
}
```
The utxo of transaction in the red block  can't be spent, so there is no need to go on .

File: rpc/watchconfirm.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		}
+		InValid := blockchain.BlockStatus(ib.GetStatus()).KnownInvalid()
+		if uint64(confirmations) >= txconf.Confirms || InValid || !isBlue {
+			ntfn := &cmds.NotificationTxConfirmNtfn{
+				ConfirmResult: cmds.TxConfirmResult{
```
Comment (lochjin) [inline rpc/watchconfirm.go:88]: I think we just need one return value (`cmds.TxConfirmResult.Confirms`) for client.  Also, these parameters should have a priority: IsValid > IsBlue > Confirms,  so why don't we set illegal transactions as `cmds.TxConfirmResult.Confirms = 0` ?  . (We purpose is just to confirm the transaction.)

**Review by lochjin — CHANGES_REQUESTED** [review]

**Review by lochjin — APPROVED** [review]

File: rpc/client/cmds/jsonrpc.go [hunk 7]
```diff
@@ -147,10 +129,28 @@ var (
 		Code:    -32603,
 		Message: "Internal error",
 	}
+	// ErrRPCDatabase indicates a database error.
+	ErrRPCDatabase = &RPCError{
+		Code:    -32604,
```
Comment (dindinw) [inline rpc/client/cmds/jsonrpc.go:134]: should follow http://xmlrpc-epi.sourceforge.net/specs/rfc.fault_codes.php and https://www.jsonrpc.org/specification 5.1 Error object

**Review by dindinw — APPROVED** [review]

**Pull request merged**